content/blog/2025-12-16-gvn-final-report.md

```diff
@@ -0,0 +1,1350 @@
++++
+title = "Optimizing Bril with Global Value Numbering: Final Report"
+[extra]
+latex = true
+bio = """
+Jake Hyun, Tobi Weinberg, and Adnan Armouti are CS PhD students at Cornell Tech.
+"""
+[[extra.authors]]
+name = "Jake Hyun"
+[[extra.authors]]
+name = "Tobi Weinberg"
+[[extra.authors]]
+name = "Adnan Armouti"
++++
+
+# Optimizing Bril with Global Value Numbering
+
+*By Jake Hyun, Tobi Weinberg, and Adnan Armouti*
+
+## Table of Contents
+
+1. [Project Goal](#1-project-goal)
+2. [Design and Implementation](#2-design-and-implementation)
+   - [2.1 GVN Algorithm](#21-gvn-algorithm)
+   - [2.2 Pruned SSA Implementation](#22-pruned-ssa-implementation)
+   - [2.3 SSA Copy Propagation](#23-ssa-copy-propagation)
+   - [2.4 Dominator-Tree Copy Elimination](#24-dominator-tree-copy-elimination)
+   - [2.5 Fixed Parallel Copy Sequentialization](#25-fixed-parallel-copy-sequentialization)
+   - [2.6 Real Phi Node Representation](#26-real-phi-node-representation)
+   - [2.7 Copy Coalescing](#27-copy-coalescing)
+   - [2.8 Phase 2 Optimizations: SSA Overhead Elimination](#28-phase-2-optimizations-ssa-overhead-elimination)
+3. [Hardest Parts to Get Right](#3-hardest-parts-to-get-right)
+4. [Empirical Evaluation](#4-empirical-evaluation)
+5. [Generative AI Usage](#5-generative-ai-usage)
+6. [Conclusion](#6-conclusion)
+7. [Comparison with Cooper/Briggs Paper](#7-comparison-with-cooperbriggs-paper)
+
+---
+
+## 1. Project Goal
+
+The goal of this project was to implement **Global Value Numbering (GVN)**, a powerful compiler optimization that eliminates redundant computations by identifying expressions that compute the same value. GVN is more powerful than local value numbering because it operates across basic block boundaries using the Static Single Assignment (SSA) form and dominator tree traversal.
+
+Specific objectives included:
+
+1. **Correctness**: The optimized programs must produce identical outputs to the original unoptimized programs for all inputs.
+2. **Effectiveness**: The optimization should reduce both static (code size) and dynamic (runtime) instruction counts.
+3. **Faithfulness to Literature**: Implement GVN using real phi nodes (as described in the original Cooper & Briggs algorithm) rather than simpler alternatives like Bril's get/set encoding.
+4. **Minimal Overhead**: The SSA conversion and lowering process should not introduce more instructions than it removes.
```

Comment:
Including this bullet raises some questions. The main goal here is GVN, but it sounds like you're implying that it also includes into-SSA and out-of-SSA transformations in the same scope? Again, I suppose I'm wondering: any particular reason to do that? Keeping things focused on just the GVN (when we're already in SSA) would keep stuff simpler, and it would help separate two sources of overhead/improvement.